File: app/src/main/java/me/zhanghai/android/files/filejob/FileJobService.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.filejob;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.IBinder;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java8.nio.file.Path;
import java9.util.function.Consumer;
import me.zhanghai.android.files.provider.common.PosixFileModeBit;
import me.zhanghai.android.files.provider.common.PosixGroup;
import me.zhanghai.android.files.provider.common.PosixUser;

public class FileJobService extends Service {

    @Nullable
    private static FileJobService sInstance;

    @NonNull
    private static final List<FileJob> sPendingJobs = new ArrayList<>();

    @NonNull
    private final List<FileJob> mRunningJobs = Collections.synchronizedList(new ArrayList<>());

    @NonNull
    private final ExecutorService mExecutorService = Executors.newCachedThreadPool();

    private FileJobNotificationManager mNotificationManager;

    @MainThread
    private static void startJob(@NonNull FileJob job, @NonNull Context context) {
        if (sInstance != null) {
            sInstance.startJob(job);
        } else {
            sPendingJobs.add(job);
            context.startService(new Intent(context, FileJobService.class));
        }
    }

    public static void archive(@NonNull List<Path> sources, @NonNull Path archiveFile,
                               @NonNull String archiveType, @Nullable String compressorType,
                               @NonNull Context context) {
        startJob(new FileJobs.Archive(sources, archiveFile, archiveType, compressorType), context);
    }

    public static void copy(@NonNull List<Path> sources, @NonNull Path targetDirectory,
                            @NonNull Context context) {
        startJob(new FileJobs.Copy(sources, targetDirectory), context);
    }

    public static void create(@NonNull Path path, boolean createDirectory,
                              @NonNull Context context) {
        startJob(new FileJobs.Create(path, createDirectory), context);
    }

    public static void delete(@NonNull List<Path> paths, @NonNull Context context) {
        startJob(new FileJobs.Delete(paths), context);
    }

    public static void move(@NonNull List<Path> sources, @NonNull Path targetDirectory,
                            @NonNull Context context) {
        startJob(new FileJobs.Move(sources, targetDirectory), context);
    }

    public static void installApk(@NonNull Path file, @NonNull Context context) {
        startJob(new FileJobs.InstallApk(file), context);
    }

    public static void open(@NonNull Path file, @NonNull String mimeType, boolean withChooser,
                            @NonNull Context context) {
        startJob(new FileJobs.Open(file, mimeType, withChooser), context);
    }

    public static void rename(@NonNull Path path, @NonNull String newName,
                              @NonNull Context context) {
        startJob(new FileJobs.Rename(path, newName), context);
    }

    public static void restoreSeLinuxContext(@NonNull Path path, boolean recursive,
                                             @NonNull Context context) {
        startJob(new FileJobs.RestoreSeLinuxContext(path, recursive), context);
    }

    public static void setGroup(@NonNull Path path, @NonNull PosixGroup group, boolean recursive,
                                @NonNull Context context) {
        startJob(new FileJobs.SetGroup(path, group, recursive), context);
    }

    public static void setMode(@NonNull Path path, @NonNull Set<PosixFileModeBit> mode,
                               boolean recursive, boolean uppercaseX, @NonNull Context context) {
        startJob(new FileJobs.SetMode(path, mode, recursive, uppercaseX), context);
    }

    public static void setOwner(@NonNull Path path, @NonNull PosixUser owner, boolean recursive,
                                @NonNull Context context) {
        startJob(new FileJobs.SetOwner(path, owner, recursive), context);
    }

    public static void setSeLinuxContext(@NonNull Path path, @NonNull String seLinuxContext,
                                         boolean recursive, @NonNull Context context) {
        startJob(new FileJobs.SetSeLinuxContext(path, seLinuxContext, recursive), context);
    }

    public static void write(@NonNull Path file, @NonNull byte[] content,
                             @Nullable Consumer<Boolean> listener, @NonNull Context context) {
        startJob(new FileJobs.Write(file, content, listener), context);
    }

    @MainThread
    public static int getRunningJobCount() {
        if (sInstance == null) {
            return 0;
        }
        return sInstance.getJobCount();
    }

    @MainThread
    public static void cancelRunningJob(int jobId) {
        if (sInstance != null) {
            sInstance.cancelJob(jobId);
        }
    }

    @Override
    public void onCreate() {
        super.onCreate();

        mNotificationManager = new FileJobNotificationManager(this);

        sInstance = this;

        Iterator<FileJob> iterator = sPendingJobs.iterator();
        while (iterator.hasNext()) {
            FileJob job = iterator.next();
            iterator.remove();
            startJob(job);
        }
    }

    private void startJob(@NonNull FileJob job) {
        mRunningJobs.add(job);
        Future<?> future = mExecutorService.submit(() -> {
            job.run(this);
            mRunningJobs.remove(job);
        });
        job.setFuture(future);
    }

    @Nullable
    @Override
    public IBinder onBind(@NonNull Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(@Nullable Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    private int getJobCount() {
        return mRunningJobs.size();
    }

    private void cancelJob(int jobId) {
        synchronized (mRunningJobs) {
            Iterator<FileJob> iterator = mRunningJobs.iterator();
            while (iterator.hasNext()) {
                FileJob job = iterator.next();
                if (job.getId() == jobId) {
                    job.cancel();
                    iterator.remove();
                    return;
                }
            }
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        sInstance = null;

        stopJobs();
        mExecutorService.shutdownNow();
    }

    private void stopJobs() {
        synchronized (mRunningJobs) {
            Iterator<FileJob> iterator = mRunningJobs.iterator();
            while (iterator.hasNext()) {
                FileJob job = iterator.next();
                job.cancel();
                iterator.remove();
            }
        }
    }

    public FileJobNotificationManager getNotificationManager() {
        return mNotificationManager;
    }
}

File: app/src/main/java/me/zhanghai/android/files/filejob/FileJobService.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.filejob

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.IBinder
import androidx.annotation.MainThread
import java8.nio.file.Path
import me.zhanghai.android.files.file.MimeType
import me.zhanghai.android.files.provider.common.PosixFileModeBit
import me.zhanghai.android.files.provider.common.PosixGroup
import me.zhanghai.android.files.provider.common.PosixUser
import me.zhanghai.android.files.util.ForegroundNotificationManager
import me.zhanghai.android.files.util.WakeWifiLock
import me.zhanghai.android.files.util.removeFirst
import java.util.concurrent.Executors
import java.util.concurrent.Future

class FileJobService : Service() {
    private lateinit var wakeWifiLock: WakeWifiLock

    internal lateinit var notificationManager: ForegroundNotificationManager
        private set

    private val executorService = Executors.newCachedThreadPool()

    private val runningJobs = mutableMapOf<FileJob, Future<*>>()

    override fun onCreate() {
        super.onCreate()

        wakeWifiLock = WakeWifiLock(FileJobService::class.java.simpleName)
        notificationManager = ForegroundNotificationManager(this)
        instance = this

        while (pendingJobs.isNotEmpty()) {
            startJob(pendingJobs.removeFirst())
        }
    }

    override fun onBind(intent: Intent): IBinder? = null

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int = START_STICKY

    private val jobCount: Int
        get() = synchronized(runningJobs) { runningJobs.size }

    private fun startJob(job: FileJob) {
        // Synchronize on runningJobs to prevent a job from removing itself before being added.
        synchronized(runningJobs) {
            val future = executorService.submit {
                job.runOn(this)
                synchronized(runningJobs) {
                    runningJobs.remove(job)
                    updateWakeWifiLockLocked()
                }
            }
            runningJobs[job] = future
            updateWakeWifiLockLocked()
        }
    }

    private fun cancelJob(id: Int) {
        synchronized(runningJobs) {
            runningJobs.removeFirst { it.key.id == id }?.value?.cancel(true)
            updateWakeWifiLockLocked()
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        instance = null

        synchronized(runningJobs) {
            while (runningJobs.isNotEmpty()) {
                runningJobs.removeFirst().value.cancel(true)
            }
            updateWakeWifiLockLocked()
        }
    }

    // Synchronize on runningJobs to avoid the potential race condition that the lock is
    // acquired after all jobs are finished in a very short time.
    private fun updateWakeWifiLockLocked() {
        wakeWifiLock.isAcquired = jobCount > 0
    }

    companion object {
        private var instance: FileJobService? = null

        private val pendingJobs = mutableListOf<FileJob>()

        val runningJobCount: Int
            @MainThread
            get() = instance?.jobCount ?: 0

        @MainThread
        private fun startJob(job: FileJob, context: Context) {
            val instance = instance
            if (instance != null) {
                instance.startJob(job)
            } else {
                pendingJobs.add(job)
                context.startService(Intent(context, FileJobService::class.java))
            }
        }

        fun archive(
            sources: List<Path>,
            archiveFile: Path,
            format: Int,
            filter: Int,
            password: String?,
            context: Context
        ) {
            startJob(ArchiveFileJob(sources, archiveFile, format, filter, password), context)
        }

        fun copy(sources: List<Path>, targetDirectory: Path, context: Context) {
            startJob(CopyFileJob(sources, targetDirectory), context)
        }

        fun create(path: Path, createDirectory: Boolean, context: Context) {
            startJob(CreateFileJob(path, createDirectory), context)
        }

        fun delete(paths: List<Path>, context: Context) {
            startJob(DeleteFileJob(paths), context)
        }

        fun move(sources: List<Path>, targetDirectory: Path, context: Context) {
            startJob(MoveFileJob(sources, targetDirectory), context)
        }

        fun installApk(file: Path, context: Context) {
            startJob(InstallApkJob(file), context)
        }

        fun open(file: Path, mimeType: MimeType, withChooser: Boolean, context: Context) {
            startJob(OpenFileJob(file, mimeType, withChooser), context)
        }

        fun rename(path: Path, newName: String, context: Context) {
            startJob(RenameFileJob(path, newName), context)
        }

        fun restoreSeLinuxContext(path: Path, recursive: Boolean, context: Context) {
            startJob(RestoreFileSeLinuxContextJob(path, recursive), context)
        }

        fun setGroup(path: Path, group: PosixGroup, recursive: Boolean, context: Context) {
            startJob(SetFileGroupJob(path, group, recursive), context)
        }

        fun setMode(
            path: Path,
            mode: Set<PosixFileModeBit>,
            recursive: Boolean,
            uppercaseX: Boolean,
            context: Context
        ) {
            startJob(SetFileModeJob(path, mode, recursive, uppercaseX), context)
        }

        fun setOwner(path: Path, owner: PosixUser, recursive: Boolean, context: Context) {
            startJob(SetFileOwnerJob(path, owner, recursive), context)
        }

        fun setSeLinuxContext(
            path: Path,
            seLinuxContext: String,
            recursive: Boolean,
            context: Context
        ) {
            startJob(SetFileSeLinuxContextJob(path, seLinuxContext, recursive), context)
        }

        fun write(
            file: Path,
            content: ByteArray,
            context: Context,
            listener: ((Boolean) -> Unit)?
        ) {
            startJob(WriteFileJob(file, content, listener), context)
        }

        @MainThread
        fun cancelJob(id: Int) {
            pendingJobs.removeFirst { it.id == id }
            instance?.cancelJob(id)
        }
    }
}
